### PR TITLE
Make week numbers in milestone and timeline sections same [WIP]

### PR DESCRIPTION
**Pull request by vaidehi44**
## What this PR does
This PR tries to resolve this issue #4621. 

Currently, the weeks in timeline section and overview tab don't show same order/number. The overview tab shows weeks which have milestone blocks. Hence, the milestones component is not correctly numbering the weeks.

The milestones component filters milestone blocks from `weeks` derived from `getWeeksArray` selector. Further, it uses `order` property to number weeks. However, the array returned by this selector only contains non-empty weeks (weeks with scheduled meetings). And, the `order` property also does not take account of this. Hence, I have defined a new selector `getAllWeeksArray` which returns an array of all the weeks including empty ones. Each week object contains keys: `empty` and `weekNumber`. If the week is non-empty, it also includes the other content being returned by `getWeeksArray` selector.

### Timeline Section: Milestone blocks defined in week 3 and 8.
week3_cropped
week8_cropped

### Milestones in overview tab:
Before: Milestone tab numbering - 2 and 5.
image

After: Milestone tab numbering - 3 and 8.
image

## Open questions and concerns
- Right now, I have just made a simple selector which groups empty and non-empty weeks in proper order. It is helping in determining if a week is empty or not, and giving its exact number through `weekNumber`. I was able to combine all these weeks using the `weekMeetings` method of `CourseDateUtils`, I am not sure if its the only correct way to do it.  Also, I think the selector can be improved and would like to get your guidance in that case.
- I am not understanding why in the first place we haven't designed a selector which would also consider empty weeks. Or, why haven't we maintained empty and non-empty weeks and their properties together.
- I am not sure if the `weekIsCompleted` method of Milestone component is working right. As you can see in the screenshot, in overview tab, it shows week 3 complete, but week 3 is still the ongoing week.

### CODE REVIEW

**Comment by ragesoss** [comment]
Looks like you may need to edit one or more of the JS tests to account for the new `allWeeks` prop.

**Comment by vaidehi44** [comment]
> Looks like you may need to edit one or more of the JS tests to account for the new `allWeeks` prop.

Yeah. Will do it.

**Comment by vaidehi44** [comment]
@ragesoss, I have tried to fix some errors raised from the build. Now, I am stuck at Ruby tests. Seems like it is raised by `timeline_rearranging_spec.rb`. Can you please help?

File: app/assets/javascripts/components/timeline/timeline.jsx [hunk 1]
```diff
@@ -271,7 +273,71 @@ const Timeline = createReactClass({
...
+              editableBlockIds={this.props.editableBlockIds}
+              edit_permissions={this.props.edit_permissions}
+              saveBlockChanges={this.props.saveBlockChanges}
+              setBlockEditable={this.props.setBlockEditable}
+              cancelBlockEditable={this.props.cancelBlockEditable}
+              updateBlock={this.props.updateBlock}
+              addBlock={this.props.addBlock}
+              deleteBlock={this.props.deleteBlock}
+              saveGlobalChanges={this.props.saveGlobalChanges}
+              canBlockMoveUp={this._canBlockMoveUp.bind(this, week, (week.order - 1))}
```
Comment (vaidehi44) [inline app/assets/javascripts/components/timeline/timeline.jsx:327]: May be, there is something wrong which I have done here. But, I can't figure it out.

**Comment by vaidehi44** [comment]
May be, there is something wrong which I have done here. But, I don't exactly know.
https://github.com/WikiEducationFoundation/WikiEduDashboard/blob/dc73305f091bf8949673a63c80a8c2782be704df/app/assets/javascripts/components/timeline/timeline.jsx#L327-L328

**Comment by vaidehi44** [comment]
Hey @ragesoss, it has passed all the checks now. The error had occurred because the numbering of `order` prop of `week`, in `timeline_rearranging_spec`, was zero-indexed while normally it starts from 1. And, in the new code, I am using `order` to decide if blocks can be rearranged or not (in `canBlockMoveUp` and `canBlockMoveDown`). Hence, I made the change in spec file. Hope it is correct.

**Comment by ragesoss** [comment]
@vaidehi44 thanks! I'm planning to have a look at this soon.

File: app/assets/javascripts/components/timeline/timeline.jsx [hunk 2]
```diff
@@ -208,6 +208,8 @@ const Timeline = createReactClass({
     // week meetings, which indicates a blackout week. Then insert the week itself.
     // The index 'i' represents the zero-index week number; both empty and non-empty
     // weeks are included in this numbering scheme.
```
Comment (ragesoss) [inline app/assets/javascripts/components/timeline/timeline.jsx:210]: These block of comments probably needs to be updated?

Comment (ragesoss) [reply app/assets/javascripts/components/timeline/timeline.jsx:210]: Actually, it probably needs to be moved to the selector.

Comment (vaidehi44) [reply app/assets/javascripts/components/timeline/timeline.jsx:210]: Okay. Will correct it right way.

**Comment by ragesoss** [comment]
I want to test this out locally and study the code a bit more, but this looks good.

I think there are some possible follow-up changes that should be done after this:
* Rather than loop through allWeeks and have an `if/else` check to insert either a Week or an EmptyWeek, this could probably now be rewritten so that every week in allWeeks renders one Week component, and if it's empty then the `Week` component is responsible for rendering an EmptyWeek instead of the its normal output.
* The number of props required for `Week` can probably be reduced. At the least, anything that is a property of the `week` or `course` props should be removable.
* the `ThisWeek` component still has its own implementation for figuring out which week is "this week", and it basically duplicates the empty week logic that got moved to the selector. It could be rewritten to use allWeeks as well.

**Comment by vaidehi44** [comment]
> I want to test this out locally and study the code a bit more, but this looks good.
> 
> I think there are some possible follow-up changes that should be done after this:
> 
> * Rather than loop through allWeeks and have an `if/else` check to insert either a Week or an EmptyWeek, this could probably now be rewritten so that every week in allWeeks renders one Week component, and if it's empty then the `Week` component is responsible for rendering an EmptyWeek instead of the its normal output.
> * The number of props required for `Week` can probably be reduced. At the least, anything that is a property of the `week` or `course` props should be removable.
> * the `ThisWeek` component still has its own implementation for figuring out which week is "this week", and it basically duplicates the empty week logic that got moved to the selector. It could be rewritten to use allWeeks as well.

So, some doubts about the follow-up changes which you have suggested:

- If we have to enable `Week` component to render empty weeks too, then `EmptyWeek` component should be entirely removed, right?
- About reducing the props in `Week` component; current props like `index`, `blocks` which can be obtained from `week` prop should be removed and, props like `timeline_start`, `timeline_end` which can be obtained from `course` prop should be removed, right? If yes, then will have to include `course` prop in `Week` as it is not present currently, right?

**Comment by ragesoss** [comment]
I'm not suggesting that these need to be done as part of this PR, just noting the areas for further improvement so I don't forget them.

> If we have to enable Week component to render empty weeks too, then EmptyWeek component should be entirely removed, right?

What I meant was that the EmptyWeek component is returned by the Week component, when appropriate.

> About reducing the props in Week component; current props like index, blocks which can be obtained from week prop should be removed and, props like timeline_start, timeline_end which can be obtained from course prop should be removed, right? If yes, then will have to include course prop in Week as it is not present currently, right?

Yes.

**Comment by vaidehi44** [comment]
@ragesoss, I am happy that it is working as expected. If you think that `allWeeks` property is fine and nothing more needs to be changed or added, then its completed from my side too.
I just feel that the `weekIsCompleted` method of Milestone component is not working right. The method indicates if a milestone week is completed or not in the overview component. Have also mentioned it in  'Open Questions and concerns'. Will confirm it again and let you know.

**Comment by ragesoss** [comment]
Ah, right. I forgot about that weekIsCompleted issue you mentioned and did not test for that. If you're having trouble figuring out why it's not behaving as expected, one good strategy could be to write a test for the expected behavior and then dig in to figure out why it doesn't match that.

**Comment by vaidehi44** [comment]
Okay. Would like to try testing. But, I don't have much idea about how to writes a test. Can you provide a starting point or an example.

**Comment by ragesoss** [comment]
In this case, since the behavior you want to test is based on the week data you pass to `Milestones` and the simple `weekIsCompleted` function, I suggest going with a plain JS test. The best starting point for an example, and possibly even the place where you want to write the test(s), is `course_date_utils.spec.js`. You can recreate the `week` and `currentWeek` values that are used in the Milestones component, and just test those in isolation. It's probably something related to how `currentWeek` gets calculated, or perhaps something related to `week.order`.

**Comment by vaidehi44** [comment]
Hi @ragesoss. 
I was able to figure out the problem with milestone's `weekIsCompleted` method. Originally, it was using `week.order` and `currentWeek` to compare the order. `currentWeek` is derived from `CourseDateUtils`'s `currentWeekOrder`. But, `currentWeekOrder` has nothing to do with `week.order` prop, as it counts the current week from `timelineStart`. Whereas, `order` prop of `week` is only applicable for non-empty weeks and won't account for the empty ones.
So, a better and simple way is to just compare it (`currentWeek`) with our new `weekNumber` property. This prop also numbers the weeks from `timelineStart` (which is the start of assignment timeline). Hence, this solves the issue.

**Comment by ragesoss** [comment]
Great. is `currentWeekOrder` calculated the same way as `week.weekNumber` in terms of whether numbering starts with 0 or 1?

**Comment by vaidehi44** [comment]
> Great. is `currentWeekOrder` calculated the same way as `week.weekNumber` in terms of whether numbering starts with 0 or 1?

Yeah. They both start with 1. `currentWeekOrder` adds 1 to `currentWeekIndex` inorder to start from 1. And, `weekNumber` is already indexed starting from 1.

**Pull request merged**